### PR TITLE
[Elm] Template improvements

### LINKED ISSUE #458: [elm-client] Update dependency: elm-date-extra
Update dependency `"justinmimbs/elm-date-extra": "2.0.0 <= v < 3.0.0"`.

There is a breaking change in the api.

```
The pattern matches things of type:

    Maybe a

But the values it will actually be trying to match are:

    Result String Date.Date
```

### See

https://github.com/OpenAPITools/openapi-generator/blob/b5942624d6714f38c0bde9ae9c7400df1d48b1ed/modules/openapi-generator/src/main/resources/elm/DateTime.mustache

https://github.com/OpenAPITools/openapi-generator/blob/b5942624d6714f38c0bde9ae9c7400df1d48b1ed/modules/openapi-generator/src/main/resources/elm/elm-package.mustache#L15

### PR DESCRIPTION
**Pull request by laughedelic**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

This PR fixes #458 and improves doc comments and line spacing in the Elm template. (see also https://github.com/swagger-api/swagger-codegen/issues/8152 for the context).

There are still some extra newlines that are hard to get rid of. They appear because the `{{>blah}}` includes bring the last newline in the parent template. I don't know if this can be somehow improved on the Mustache side, but in general it's quite hard to produce some code that immediately complies with the formatting guideline (elm-format).

### CODE REVIEW

**Comment by laughedelic** [comment]
e4fa432 also solves https://github.com/swagger-api/swagger-codegen/issues/8151

**Comment by wing328** [comment]
> They appear because the {{>blah}} includes bring the last newline in the parent template.

@laughedelic thanks for the PR 👍 

Would removing the last newline in the parent (partial) template help? If you can point me to the line number in the template and show me the actual vs expected output, I can help take a look.

cc @trenneman

**Comment by laughedelic** [comment]
> Would removing the last newline in the parent (partial) template help? If you can point me to the line number in the template and show me the actual vs expected output, I can help take a look.

As I understand it, the extra newline comes from the child template. So if you have an `{{>child}}` at the last line of the parent template, the result will have `\n\n` at the end, coming from the parent and the child.

Here is an example piece of template:

https://github.com/OpenAPITools/openapi-generator/blob/e4fa4329a71a7538fed5dbfc67c38fe937fd5a1f/modules/openapi-generator/src/main/resources/elm/modelTypeAlias.mustache#L13-L16

And the produced result:

https://github.com/OpenAPITools/openapi-generator/blob/e4fa4329a71a7538fed5dbfc67c38fe937fd5a1f/samples/client/petstore/elm/src/Data/ApiResponse.elm#L30-L47

One extra newline is added after inlining the `aliasDecoder`, same with the `aliasEncoder`, and then one more in the end, because the `modelTypeAlias` is itself inlined in the `model` template.

File: bin/elm-petstore.sh [hunk 1]
```diff
@@ -27,6 +27,6 @@ fi
 
 # if you've executed sbt assembly previously it will use that instead.
 export JAVA_OPTS="${JAVA_OPTS} -XX:MaxPermSize=256M -Xmx1024M -DloggerPath=conf/log4j.properties"
-ags="generate -i modules/openapi-generator/src/test/resources/2_0/petstore.yaml -g elm -o samples/client/petstore/elm $@"
+ags="generate -i modules/openapi-generator/src/test/resources/2_0/petstore.yaml -g elm -t modules/openapi-generator/src/main/resources/elm -o samples/client/petstore/elm $@"
```
Comment (trenneman) [inline bin/elm-petstore.sh]: do we need this?

File: modules/openapi-generator/src/main/resources/elm/DateOnly.mustache [hunk 2]
```diff
@@ -24,11 +25,8 @@ dateOnlyEncoder model =
...
+        Result.Ok date ->
             Decode.succeed date
 
-        Nothing ->
-            Decode.fail <|
-                "Cannot convert "
-                    ++ str
-                    ++ " to DateOnly"
+        Result.Err msg ->
+            Decode.fail msg
```
Comment (trenneman) [inline modules/openapi-generator/src/main/resources/elm/DateOnly.mustache]: Something like `Decode.fail <| "Cannot decode ISO string: " ++ msg` makes clear that it's the decoder that failed

File: modules/openapi-generator/src/main/resources/elm/DateTime.mustache [hunk 3]
```diff
@@ -24,11 +25,8 @@ dateTimeEncoder model =
...
+        Result.Ok date ->
             Decode.succeed date
 
-        Nothing ->
-            Decode.fail <|
-                "Cannot convert "
-                    ++ str
-                    ++ " to DateTime"
+        Result.Err msg ->
+            Decode.fail msg
```
Comment (trenneman) [inline modules/openapi-generator/src/main/resources/elm/DateTime.mustache]: The same here

File: bin/elm-petstore.sh (hunk 1 above)
Comment (wing328) [reply bin/elm-petstore.sh]: Yes, I think so as using -t allows us to regenerate the samples without doing another "mvn clean package" after updating the templates.

File: modules/openapi-generator/src/main/resources/elm/DateOnly.mustache (hunk 2 above)
Comment (laughedelic) [reply modules/openapi-generator/src/main/resources/elm/DateOnly.mustache]: The error message in the new date-extra is

> Failed to create a Date from string 'foo': Invalid ISO 8601 format

Do you still want to add `"Cannot decode ISO string: "`?

File: bin/elm-petstore.sh (hunk 1 above)
Comment (laughedelic) [reply bin/elm-petstore.sh]: Yes, I also checked that it's used in other scripts.

Comment (trenneman) [reply bin/elm-petstore.sh]: It feels a bit weird that it sort of overloads the build, but it makes sense to add it if that's also the case for the other generators.

File: modules/openapi-generator/src/main/resources/elm/DateOnly.mustache (hunk 2 above)
Comment (trenneman) [reply modules/openapi-generator/src/main/resources/elm/DateOnly.mustache]: Well the thing is this is a `Decoder` error, not a `Date` error. But the type already indicates that so I think we can stick with the `Decode.fail msg`.

**Comment by laughedelic** [comment]
Hi! I replied to the review comments and now I'm not sure if there's still anything left to change.

**Comment by wing328** [comment]
If no more question/feedback on this PR, I'll merge it into master tomorrow (Monday)

**Comment by andys8** [comment]
Nice!

**Comment by wing328** [comment]
Hi all, I've filed https://github.com/OpenAPITools/openapi-generator/pull/879 to improve the code format of the Elm client by removing the EOF from 2 mustache files. Please review to see if it looks good to you.

**Comment by wing328** [comment]
#879 has been reviewed and merged into master.

Shall we use https://github.com/jfmengels/elm-lint to improve the auto-generated Elm code?

**Comment by trenneman** [comment]
@wing328 You typically use `elm-format` for that. You want to add this as a post-processing step?

**Comment by andys8** [comment]
`elm-format` does formatting and minor changes (like changing import order). `elm-analyse` is a longer (static code analysis). We could avoid the two dependencies by making sure the templates themselves are formatted correctly. But it could also be a benefit to make it sure by using `elm-format`.

**Comment by wing328** [comment]
Thanks for the suggestion guys. I will see if I can come up with a solution to run `elm-format` for each auto-generated Elm files. Will keep you guys posted.

**Comment by trenneman** [comment]
That would help. However, my preference would be to format it correctly from the start, without any post-processing.

**Comment by wing328** [comment]
> That would help. However, my preference would be to format it correctly from the start, without any post-processing.

Totally agree but we found that it's not easy to do so with mustache.

For Go, we're trying to use `gofmt -w` and the result looks good so far: https://github.com/OpenAPITools/openapi-generator/pull/929

I've created https://github.com/OpenAPITools/openapi-generator/issues/933 to track the code format improvement

**Pull request merged**